miniconf/src/packed.rs

```diff
@@ -0,0 +1,219 @@
+        value.0
+    }
+}
+
+impl Deref for Packed {
+    type Target = NonZeroUsize;
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for Packed {
+    #[inline]
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl Packed {
+    /// Number of bits in the representation including the marker bit
+    pub const BITS: u32 = NonZeroUsize::BITS;
+
+    /// The total number of bits this representation can store.
+    pub const CAPACITY: u32 = Self::BITS - 1;
+
+    /// The empty value
+    pub const EMPTY: Self = Self(
+        // Slightly cumbersome to generate it with `const`
+        NonZeroUsize::MIN
+            .saturating_add(1)
+            .saturating_pow(Self::CAPACITY),
+    );
+
+    /// Create a new `Packed` from a `usize`.
+    ///
+    /// The value must not be zero.
+    #[inline]
+    pub const fn new(value: usize) -> Option<Self> {
+        match NonZeroUsize::new(value) {
+            Some(value) => Some(Self(value)),
+            None => None,
+        }
+    }
+
+    /// The value is empty.
+    #[inline]
+    pub const fn is_empty(&self) -> bool {
+        matches!(*self, Self::EMPTY)
+    }
+
+    /// Clear and discard all bits stored.
+    #[inline]
+    pub fn clear(&mut self) {
+        *self = Self::EMPTY;
+    }
+
+    /// Number of bits stored.
+    #[inline]
+    pub const fn len(&self) -> u32 {
+        Self::CAPACITY - self.0.trailing_zeros()
+    }
+
+    /// Return the representation aligned to the LSB with the marker bit
+    /// moved from the LSB to the MSB.
+    #[inline]
+    pub fn into_lsb(&self) -> NonZeroUsize {
+        // Note(unwrap): we ensure there is at least the marker bit set
+        NonZeroUsize::new(((self.get() >> 1) | (1 << Self::CAPACITY)) >> self.trailing_zeros())
+            .unwrap()
+    }
+
+    /// Build a `Packed` from a LSB-aligned representation with the marker bit
+    /// moved from the MSB the LSB.
+    #[inline]
+    pub fn from_lsb(value: NonZeroUsize) -> Self {
+        // Note(unwrap): we ensure there is at least the marker bit set
+        Self::new(((value.get() << 1) | 1) << value.leading_zeros()).unwrap()
+    }
+
+    /// Return the number of bits required to represent `num`.
+    ///
+    /// Ensures that at least one bit is allocated.
+    #[inline]
+    pub fn bits_for(num: usize) -> u32 {
+        (Self::BITS - num.leading_zeros()).max(1)
+    }
+
+    /// Remove the given number of MSBs and return them.
+    ///
+    /// If the value does not contain sufficient bits
+    /// it is left unchanged and `None` is returned.
+    ///
+    /// # Args
+    /// * `bits`: Number of bits to pop. `bits <= Self::CAPACITY`
+    #[inline]
+    pub fn pop_msb(&mut self, bits: u32) -> Option<usize> {
+        let s = self.get();
+        if let Some(new) = Self::new(s << bits) {
+            self.0 = new.0;
+            Some((s >> (Self::CAPACITY - bits)) >> 1)
```

Comment:
Maybe add a note that the extra >>1 is for getting rid of the marker bit. I didn't follow this at first.

Reply:
It's to avoid shift wrap. Comment added.